text/0113-secure-script-container.md

```diff
@@ -0,0 +1,183 @@
+
+A syntax like the following could be provided:
+
+```html
+<secure-script extra-globals="ga"></secure-script>
+```
+
+### Known Limitations
+
+* Currently there is an outstanding issue with synthetic shadow in that the Event object patching happens lazily, so we are unable to prevent its patching at the time of creating the secure container, which means that events are still retargeted. [See LWC PR #1569](https://github.com/salesforce/lwc/pull/1569).
+
+
+### Testing
+
+We will want to gather the top 20 common libraries and implement their behavior inside of `secure-script` to ensure that we are prooperly accounting for all these libraries' use cases. This will be our integration test suite. 
+
+### Prototype Code
+
+```js
+import createSecureEnvironment from '@caridy/sjs/lib/browser-realm';
+
+// creating the new sandbox with all global endowments
+const evaluate = createSecureEnvironment(
+    undefined,
+    window,
+);
+
+evaluate(`
+    // This initialization will prevent any of these APIs to be polyfilled
+    // on the blue realm that can affect this sandbox.
+    const pE = Object.preventExtensions;
+    pE(HTMLElement.prototype);
+    pE(Element.prototype);
+    pE(Node.prototype);
+    pE(Event.prototype);
+    pE(Document.prototype);
+    pE(EventTarget.prototype);
+    pE(MutationObserver.prototype);
+    pE(HTMLCollection.prototype);
+    pE(NodeList.prototype);
+    pE(ShadowRoot.prototype);
+    pE(HTMLSlotElement.prototype);
+    pE(Text.prototype);
+`);
+
+class SecureScript extends HTMLElement {
+    connectedCallback() {
+        const src = this.textContent;
+        if (src) {
+            evaluate(src);
+        }
+    }
+}
+
+customElements.define('secure-script', SecureScript);
+```
+
+## Drawbacks
+
+- We are potentially prolonging the life of our synthetic shadow dom polyfill, instead of encouraging community library authors to adopt shadow DOM semantics.
+- The element may end up being anywhere from 5kb - 20kb.
+
+## Alternatives
+
+- **Allow shadow DOM to be toggled at a more granular level.** There is a forthcoming RFC for allowing component authors to decide if they want to opt in to shadow DOM or not. This would mean we can keep top-level elements and layout components in the light DOM, but base components or managed components can choose to safeguard their internals. Until we allow this, `secure-script` serves as a mitigation for application developers looking to add third party libraries to their apps.
+- Do not apply the shadow DOM polyfill at all: this would allow component authors to violate shadow DOM encapsulation which is not an option.
+- If we don't do this, then companies / websites building on our platform will not be able to integrate with common third party libraries.
+
+## Adoption strategy
+
+TBD
```

Comment:
An application developer can choose to expose it or not, so I think it could be both. For Salesforce Communities / Commerce we would probably do the latter since we side on giving the admin the markup.

Reply:
@gregwhitworth you said you had more opinions here?

Reply:
I'm hesitant to allow this to be a free-for-all due to this statement in your RFC

> We are potentially prolonging the life of our synthetic shadow dom polyfill, instead of encouraging community library authors to adopt Shadow DOM semantics.

I would prefer for us to leverage the top libs identified and give them access to this for Phase 1 and when the need for synthetic shadow is removed we can update the <n> libs to leverage Phase 2.

Is this a large request outside of these large products?

Reply:
> we can update the libs

That part is throwing me off a bit. The only work on us will be to support a native shadow implementation of `virtual-script` but this will be transparent for libraries. I'm certainly misunderstanding something.

Reply:
Thanks, and sorry for the delayed response. We've since discussed this further but it comes down to my concern of compat risk, eg: who can use this and when? If we allow anyone to utilize it and run any code they want to bypass the shadow then we need to never close that hole. That puts a technical debt on the LWC team to maintain that and it may come with a whole host of other issues such as other teams that desire the encapsulation (for the very reason it exists) and them not being able to evolve due to this script. So ultimately I'd like to understand this plan further and given the business scenarios I'm aware of I think there can be a path forward that not only blocks us from a compat/tech debt perspective but also in the prodsec arena as well.

Reply:
@gregwhitworth is it fair to say that the usage of `virtual-script` is outside of the scope of this RFC? I can certainly flesh that out in the "Rollout strategy". Something to the tune of "usage of virtual script containers require above average scrutiny because of a, b, c..."?

Reply:
We discussed this offline and agreed to something like the above.